_drafts/ember-orbit.md

```diff
@@ -0,0 +1,477 @@
+
+- store beforeQuery
+- store query
+
+When you add a message, you'll see a number of more entries:
+
+- store beforeUpdate
+- remote beforePush
+- remote transform
+- remote push
+- store transform
+- store update
+
+DETAILS ON WHAT EACH OF THESE MEANS SPECIFICALLY
+
+## Pull and Sync: Querying Data from the Server
+
+We're saving data to the server, but if you reload your app you'll notice we aren't loading the data yet; we simply query the in-memory store. What we want to do is, every time the in-memory store is queried, we want to query the remote store first.
+
+`app/data-strategies/remote-request.js`:
+
+```js
+import { RequestStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new RequestStrategy({
+      name: 'remote-request',
+
+      source: 'store',
+      on: 'beforeQuery',
+
+      target: 'remote',
+      action: 'pull',
+
+      blocking: true,
+    });
+  }
+};
+```
+
+This says that before the local store is queried, pull from the remote store.
+
+If you reload your app, this isn't enough to get the data load working. We also need to sync the data. The data is pulled down from the server, but it's left in limbo; it isn't transferred into our store. CONFIRM THIS EXPLANATION IS CORRECT. We need a sync strategy to do that:
+
+`app/data-strategies/remote-sync.js`:
+
+```js
+import { SyncStrategy } from '@orbit/coordinator';
+
+export default {
+  create() {
+    return new SyncStrategy({
+      name: 'remote-sync',
+
+      source: 'remote',
+      target: 'store',
+
+      blocking: false,
+    });
+  }
+};
+```
+
+Now if we reload our app we'll see all the data from the server appear upon load. When we add messages, they're added both to our local store and to the remote API.
+
+Now we've replicated the functionality of Ember Data. It took some work to get there! But now’s where that work pays off; we can start adding offline storage and synchronization features.
+
+## IndexedDBSource: Persisting Data Locally
+
+Right now our app requires us to be online; if the API server isn't reachable, the app won't even launch. To confirm this, stop the Rails API server by pressing control-C in the terminal that's running the Rails server. Reload your Ember app. You'll see a blank white screen and an error in the console.
+
+SCREENSHOT
+
+If we backed up our data locally in the browser, we would be able to load it even if the server wasn't reachable. (We could even use [the `ember-service-worker` addon](https://github.com/DockYard/ember-service-worker/) to allow our static assets to be loadable even when the device has no internet connection at all. We won’t get into that in this post, but give it a try on your own!)
+
+Ember-Orbit will allow us to store our data locally like that, using the IndexedDB browser technology. (There is also a Local Storage option you can fall back to. WHICH BROWSERS WOULD THIS BE USEFUL FOR?) We installed the NPM Orbit packages to support IndexedDB earlier.
```

Comment:
I couldn't find any browsers in caniuse.com that currently support Local Storage but not IndexedDB. Is there any reason Local Storage would still be beneficial to include?

Reply:
At this point, there aren't any evergreen browsers that support local storage but not IndexedDB. IE 10 & 11 only have partial support for IndexedDB and thus could fall back to local storage if needed. Luckily, there are fewer calls to support these browsers now.